File: web/src/types/policy.ts
// 政策相关类型定义

// 来源文档类型
export interface SourceDocument {
  title?: string
  doc_number?: string  // 官方文号（非必填）
  url: string
  extracted_text?: string
}

// ── 主政策类型 ──────────────────────────────────────────────

export interface Policy {
  policy_id: string
  policy_type: string
  title: string
  region_code: string
  region_name?: string

  source_attachments?: string  // JSON string of SourceDocument[]

  published_at: string
  effective_start: string
  effective_end?: string
  policy_year?: number

  status: string
  version: number

  // 通用类型扩展数据
  type_data?: Record<string, any>

  // 向后兼容
  social_insurance?: PolicySocialInsurance

  created_at: string
  updated_at: string
  created_by?: string
  reviewed_by?: string
}

// 政策列表项（API返回的扁平化结构）
export interface PolicyListItem {
  policy_id: string
  policy_type: string
  title: string
  region_code: string
  region_name?: string

  // 类型摘要
  type_summary?: Record<string, any>

  // 向后兼容（社保类型）
  si_upper_limit?: number
  si_lower_limit?: number

  effective_start: string
  status: string
  is_retroactive: boolean
}

// ── 社保基数类型 ──────────────────────────────────────

export interface PolicySocialInsurance {
  si_upper_limit?: number
  si_lower_limit?: number
  si_avg_salary_ref?: number
  is_retroactive: boolean
  retroactive_start?: string
  retroactive_months?: number
  coverage_types: string[]
  change_rate_upper?: number
  change_rate_lower?: number
  special_notes?: string
}

export interface PolicyHousingFund {
  hf_upper_limit?: number
  hf_lower_limit?: number
  is_retroactive: boolean
  retroactive_start?: string
  retroactive_months?: number
  change_rate_upper?: number
  change_rate_lower?: number
  special_notes?: string
}

// ── 创建/更新 ────────────────────────────────────────────────

export interface PolicyCreateInput {
  policy_type?: string
  title: string
  region_code: string
  published_at: string
  effective_start: string
  effective_end?: string

  // 通用类型扩展数据入口
  type_data?: Record<string, any>

  // 向后兼容
  social_insurance?: {
    si_upper_limit?: number
    si_lower_limit?: number
    is_retroactive?: boolean
    retroactive_start?: string
    coverage_types?: string[]
    special_notes?: string
  }
  raw_content?: string
}

export interface PolicyUpdateInput {
  title?: string
  published_at?: string
  effective_start?: string
  effective_end?: string

  type_data?: Record<string, any>

  // 向后兼容
  social_insurance?: Partial<PolicyCreateInput['social_insurance']>

  change_reason: string
  create_new_version?: boolean
  special_notes?: string
}

export interface Region {
  code: string
  name: string
  level: 'country' | 'province' | 'city' | 'district'
  parent_code?: string
  full_path?: string
  min_wage?: number
  avg_salary?: number
}

// ── 状态枚举 ────────────────────────────────────────────────

export const POLICY_STATUS_MAP: Record<string, { label: string; color: string }> = {
  draft: { label: '草稿', color: 'default' },
  pending_review: { label: '待审核', color: 'orange' },
  active: { label: '生效中', color: 'green' },
  expired: { label: '已过期', color: 'red' },
  revoked: { label: '已撤销', color: 'default' },
}

export const REVIEW_STATUS_MAP: Record<string, { label: string; color: string }> = {
  pending: { label: '待审核', color: 'orange' },
  claimed: { label: '审核中', color: 'blue' },
  approved: { label: '已通过', color: 'green' },
  rejected: { label: '已拒绝', color: 'red' },
  needs_clarification: { label: '需补充', color: 'purple' },
}

export const PRIORITY_MAP: Record<string, { label: string; color: string }> = {
  urgent: { label: '紧急', color: 'red' },
  high: { label: '高', color: 'orange' },
  normal: { label: '普通', color: 'blue' },
  low: { label: '低', color: 'default' },
}

export const RISK_LEVEL_MAP: Record<string, { label: string; color: string }> = {
  high: { label: '高风险', color: 'red' },
  medium: { label: '中风险', color: 'orange' },
  low: { label: '低风险', color: 'green' },
}

// 政策类型映射
export const POLICY_TYPE_MAP: Record<string, { label: string; color: string }> = {
  social_insurance: { label: '社保基数', color: 'blue' },
  social_insurance_base: { label: '社保基数', color: 'blue' },  // 兼容旧数据
  housing_fund: { label: '公积金基数', color: 'geekblue' },
  avg_salary: { label: '社会平均工资', color: 'cyan' },
  talent_policy: { label: '人才政策', color: 'purple' },
}

// 获取政策类型标签（兼容动态类型）
export function getPolicyTypeLabel(typeCode?: string): { label: string; color: string } {
  if (!typeCode) return { label: '未知', color: 'default' }
  return POLICY_TYPE_MAP[typeCode] || { label: typeCode, color: 'default' }
}

File: app/models/policy.py
"""政策模型"""
from sqlalchemy import Column, Text, Integer, ForeignKey, Index
from datetime import datetime
from app.database import Base
import uuid


class Policy(Base):
    """政策主表"""
    __tablename__ = "policies"

    policy_id = Column(Text, primary_key=True, default=lambda: str(uuid.uuid4()))
    policy_type = Column(Text, default="social_insurance")
    title = Column(Text, nullable=False)
    region_code = Column(Text, nullable=False)

    source_attachments = Column(Text, default="[]")

    published_at = Column(Text, nullable=False)
    effective_start = Column(Text, nullable=False)
    effective_end = Column(Text)
    policy_year = Column(Integer)

    status = Column(Text, default="draft")
    version = Column(Integer, default=1)

    # 动态类型的扩展数据（JSON 存储）
    # 仅用于通过管理后台动态创建的政策类型（无专用扩展表的类型）
    extension_data = Column(Text, default=None)

    raw_content = Column(Text)
    raw_snapshot_url = Column(Text)

    created_at = Column(Text, default=lambda: datetime.utcnow().isoformat())
    updated_at = Column(Text, default=lambda: datetime.utcnow().isoformat())
    created_by = Column(Text)
    reviewed_by = Column(Text)
    reviewed_at = Column(Text)

    deleted_at = Column(Text)
    deleted_by = Column(Text)



# Indexes for common query patterns
Index('ix_policies_region_type_start', Policy.region_code, Policy.policy_type, Policy.effective_start)
Index('ix_policies_status_deleted', Policy.status, Policy.deleted_at)
Index('ix_policies_policy_year', Policy.policy_year)


class PolicySocialInsurance(Base):
    """社保基数扩展表"""
    __tablename__ = "policy_social_insurance"

    policy_id = Column(Text, ForeignKey("policies.policy_id"), primary_key=True)

    si_upper_limit = Column(Integer)
    si_lower_limit = Column(Integer)
    si_avg_salary_ref = Column(Integer)

    is_retroactive = Column(Integer, default=0)
    retroactive_start = Column(Text)
    retroactive_months = Column(Integer)

    coverage_types = Column(Text, default='["养老", "医疗", "失业", "工伤", "生育"]')

    prev_si_upper = Column(Integer)
    prev_si_lower = Column(Integer)
    change_rate_upper = Column(Text)
    change_rate_lower = Column(Text)

    special_notes = Column(Text)


class PolicyHousingFund(Base):
    """公积金基数扩展表"""
    __tablename__ = "policy_housing_fund"

    policy_id = Column(Text, ForeignKey("policies.policy_id"), primary_key=True)

    hf_upper_limit = Column(Integer)
    hf_lower_limit = Column(Integer)

    is_retroactive = Column(Integer, default=0)
    retroactive_start = Column(Text)
    retroactive_months = Column(Integer)

    prev_hf_upper = Column(Integer)
    prev_hf_lower = Column(Integer)
    change_rate_upper = Column(Text)
    change_rate_lower = Column(Text)

    special_notes = Column(Text)
